### PR TITLE
[Validator] Add `MacAddress` constraint for validating MAC address

### PR DESCRIPTION
**Pull request by Ninos**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| License       | MIT

Possibility to validate against mac address.

See also past discussion: #51777

### CODE REVIEW

**Comment by OskarStark** [comment]
What about renaming it to `MacAddress`?

File: src/Symfony/Component/Validator/Tests/Constraints/MacValidatorTest.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+     * @dataProvider getValidMacs
+     */
+    public function testValidMac($mac)
+    {
+        $this->validator->validate($mac, new Mac());
+
+        $this->assertNoViolation();
+    }
+
+    public static function getValidMacs()
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/Tests/Constraints/MacValidatorTest.php]:
```suggestion
    public static function getValidMacs(): array
```

File: src/Symfony/Component/Validator/Tests/Constraints/MacValidatorTest.php [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    public function testValidMacsWithWhitespaces($mac)
+    {
+        $this->validator->validate($mac, new Mac([
+            'normalizer' => 'trim',
+        ]));
+
+        $this->assertNoViolation();
+    }
+
+    public static function getValidMacsWithWhitespaces()
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/Tests/Constraints/MacValidatorTest.php]:
```suggestion
    public static function getValidMacsWithWhitespaces(): array
```

File: src/Symfony/Component/Validator/Tests/Constraints/MacValidatorTest.php [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+        $this->validator->validate($mac, $constraint);
+
+        $this->buildViolation('myMessage')
+            ->setParameter('{{ value }}', '"'.$mac.'"')
+            ->setCode(Mac::INVALID_MAC_ERROR)
+            ->assertRaised();
+    }
+
+    public static function getInvalidMacs()
```
Comment (OskarStark) [inline src/Symfony/Component/Validator/Tests/Constraints/MacValidatorTest.php]:
```suggestion
    public static function getInvalidMacs(): array
```

**Comment by Ninos** [comment]
> What about renaming it to `MacAddress`?

For me `Mac` was cleaner, because the `Ip` constraint is not called `IpAddress` (consistency) & there are no different meanings of the word `mac`, so no conflicts

**Comment by derrabus** [comment]
> there are no different meanings of the word `mac`

I see what you did there. 🍏

**Comment by geekabel** [comment]
> What about renaming it to `MacAddress`?

I actually find this name more relevant than simply putting `Mac`

**Comment by Ninos** [comment]
> I see what you did there. 🍏

Hahah, seems I'm so "anti Apple", that I forgot one of the most known names in world... :D

I don't think there will be any constraint used in real scenario to check mac devices (mac device names?) - but if, the constraint should be a prefixed with the name of the company, e.g. `AppleMac`. But I can change it to `MacAddress`, if preferred, just let me know :-)

**Comment by xabbuh** [comment]
As I said in https://github.com/symfony/symfony/pull/51777#issuecomment-1742578118 I am not convinced that this is a common enough use case to be part of the Symfony core.

**Comment by derrabus** [comment]
> I am not convinced that this is a common enough use case to be part of the Symfony core.

Yeah, maybe we should encourage releasing niche-y validators as standalone packages more.

Then again, we have other network-related constraints, like the already mentioned IP and CIDR. What makes this one different?

**Comment by OskarStark** [comment]
To me a `MacAddress` is general enough to be part of the core.

I vote for naming it `MacAddress`.

**Comment by xabbuh** [comment]
> Then again, we have other network-related constraints, like the already mentioned IP and CIDR. What makes this one different?

I can work with a client's IP address. But for MAC addresses I do only get the one from the latest hop.

**Comment by OskarStark** [comment]
@Ninos can you please elaborate on your use-case? Thanks

**Comment by geekabel** [comment]
> To me a `MacAddress` is general enough to be part of the core.
> 
> I vote for naming it `MacAddress`.

I vote `MacAddress` too

**Comment by Ninos** [comment]
Renamed to `MacAddress`, hope it's fine now :-)

> can you please elaborate on your use-case?

Application e.g. for storing device specific mac addresses

File: src/Symfony/Component/Validator/CHANGELOG.md [hunk 4]
```diff
@@ -4,6 +4,7 @@ CHANGELOG
 6.4
 ---
 
+ * Add `MacAddress` constraint
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/CHANGELOG.md:7]: should be moved to a new 7.1 section now

File: src/Symfony/Component/Validator/Constraints/MacAddress.php [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+namespace Symfony\Component\Validator\Constraints;
+
+use Symfony\Component\Validator\Constraint;
+use Symfony\Component\Validator\Exception\InvalidArgumentException;
+
+/**
+ * Validates that a value is a valid MAC address.
+ *
+ * @Annotation
+ * @Target({"PROPERTY", "METHOD", "ANNOTATION"})
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/MacAddress.php]: we don't support annotations anymore \o/

File: src/Symfony/Component/Validator/Constraints/MacAddress.php [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+#[\Attribute(\Attribute::TARGET_PROPERTY | \Attribute::TARGET_METHOD | \Attribute::IS_REPEATABLE)]
+class MacAddress extends Constraint
+{
+    public const INVALID_MAC_ERROR = 'a183fbff-6968-43b4-82a2-cc5cf7150036';
+
+    protected const ERROR_NAMES = [
+        self::INVALID_MAC_ERROR => 'INVALID_MAC_ERROR',
+    ];
+
+    public string $message = 'This is not a valid MAC address.';
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/MacAddress.php:31]: this message should be added to the corresponding translation file I suppose

Comment (Ninos) [reply src/Symfony/Component/Validator/Constraints/MacAddress.php:31]: How can I regenerate the translation files? :angel:

Comment (nicolas-grekas) [reply src/Symfony/Component/Validator/Constraints/MacAddress.php:31]: you should add it to the `src/Symfony/Component/Validator/Resources/translations/validators.en.xlf` file on branch 5.4 (and to the translation files in the languages you know at the same time)

File: src/Symfony/Component/Validator/CHANGELOG.md [hunk 7]
```diff
@@ -1,6 +1,11 @@
 CHANGELOG
 =========
 
+7.1
+---
+
+* Add `MacAddress` constraint
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/CHANGELOG.md]:
```suggestion
 * Add `MacAddress` constraint
```

**Review by nicolas-grekas — APPROVED** [review]

File: src/Symfony/Component/Validator/Resources/translations/validators.en.xlf [hunk 8]
```diff
@@ -426,6 +426,10 @@
                 <source>Using hidden overlay characters is not allowed.</source>
                 <target>Using hidden overlay characters is not allowed.</target>
             </trans-unit>
+            <trans-unit id="110">
```
Comment (fabpot) [inline src/Symfony/Component/Validator/Resources/translations/validators.en.xlf]:
```suggestion
            <trans-unit id="111">
```

As we already have 110 (you will get it if you rebase on current 7.1).

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Validator/Resources/translations/validators.en.xlf (hunk 8 above)
Comment (OskarStark) [reply src/Symfony/Component/Validator/Resources/translations/validators.en.xlf]: Done, we are already at 112

**Comment by OskarStark** [comment]
PR rebased ✅

**Review by nicolas-grekas — APPROVED** [review]
Still GTM. Let's make a decision on this one?

**Review by OskarStark — APPROVED** [review]

**Comment by xabbuh** [comment]
I still don’t see this as a common enough use case, but I won’t insist if others agree with adding this constraint.

**Comment by nicolas-grekas** [comment]
Thank you @Ninos.

**Pull request merged**